src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/base/Functions.java

```diff
@@ -396,19 +499,41 @@ public static boolean checkBuiltInFunctionExists(FunctionIdentifier identifier)
 
     public static RuntimeIterator getBuiltInFunctionIterator(
             FunctionIdentifier identifier,
-            IteratorMetadata metadata,
-            List<RuntimeIterator> arguments
+            List<RuntimeIterator> arguments,
+            IteratorMetadata metadata
     ) {
-        Class<? extends RuntimeIterator> functionClass = builtInFunctions.get(identifier);
-        try {
-            Constructor<? extends RuntimeIterator> ctor = functionClass.getConstructor(
-                List.class,
-                IteratorMetadata.class
+        BuiltinFunction builtinFunction = builtInFunctions.get(identifier);
+
+        for (int i = 0; i < arguments.size(); i++) {
+            arguments.set(
+                i,
+                new TypePromotionIterator(
+                        "Invalid argument for function " + identifier.getName() + ". ",
+                        arguments.get(i),
+                        builtinFunction.getSignature().getParameterTypes().get(i),
+                        metadata
+                )
             );
-            return ctor.newInstance(arguments, metadata);
-        } catch (Exception e) {
-            throw new RuntimeException(e.getMessage());
         }
+
+        RuntimeIterator functionCallIterator;
+        try {
+            Constructor<? extends RuntimeIterator> constructor = builtinFunction.getFunctionIterator()
+                .getConstructor(
+                    List.class,
+                    IteratorMetadata.class
+                );
+            functionCallIterator = constructor.newInstance(arguments, metadata);
+        } catch (ReflectiveOperationException e) {
+            throw new UnknownFunctionCallException(identifier.getName(), arguments.size(), metadata);
+        }
+
+        return new TypePromotionIterator(
```

Comment:
And likewise, after disentangling (see my other comment), this can become:

new TreatAsIterator(
    ...
    new TypePromotionIterator(
        ...
        functionCallIterator
    ...
)

where the iterators cleanly call each other without even being aware of each other, and without any encapsulation breach.